src/main/java/com/google/swarm/tokenization/common/WriteToGCS.java

```diff
@@ -0,0 +1,112 @@
+package com.google.swarm.tokenization.common;
+
+import com.google.auto.value.AutoValue;
+import com.google.privacy.dlp.v2.Table;
+import org.apache.beam.sdk.coders.StringUtf8Coder;
+import org.apache.beam.sdk.io.Compression;
+import org.apache.beam.sdk.io.FileIO;
+import org.apache.beam.sdk.io.TextIO;
+import org.apache.beam.sdk.io.WriteFilesResult;
+import org.apache.beam.sdk.transforms.Contextful;
+import org.apache.beam.sdk.transforms.DoFn;
+import org.apache.beam.sdk.transforms.PTransform;
+import org.apache.beam.sdk.transforms.ParDo;
+import org.apache.beam.sdk.transforms.windowing.BoundedWindow;
+import org.apache.beam.sdk.transforms.windowing.PaneInfo;
+import org.apache.beam.sdk.values.KV;
+import org.apache.beam.sdk.values.PCollection;
+import org.checkerframework.checker.initialization.qual.Initialized;
+import org.checkerframework.checker.nullness.qual.NonNull;
+import org.checkerframework.checker.nullness.qual.UnknownKeyFor;
+
+import java.text.DecimalFormat;
+import java.util.List;
+import java.util.stream.Collectors;
+
+@AutoValue
+@SuppressWarnings("serial")
+public abstract class WriteToGCS extends PTransform<PCollection<KV<String, Table.Row>>, WriteFilesResult<String>> {
+
+    public abstract String outputBucket();
+
+    public abstract Util.FileType fileType();
+
+    public static Builder newBuilder() {
+        return new AutoValue_WriteToGCS.Builder();
+    }
+
+    @AutoValue.Builder
+    public abstract static class Builder {
+        public abstract Builder setOutputBucket(String value);
+
+        public abstract Builder setFileType(Util.FileType value);
+
+        public abstract WriteToGCS build();
+    }
+
+    @Override
+    public WriteFilesResult<String> expand(PCollection<KV<String, Table.Row>> input) {
+
+        return input.apply("ConvertTableRowToString",ParDo.of(new ConvertTableRowToString(fileType())))
+                .apply("WriteFileToGCS", FileIO.<String, KV<String, String>>writeDynamic()
+                    .by(KV::getKey)
+                    .withDestinationCoder(StringUtf8Coder.of())
+                    .via(Contextful.fn(KV::getValue), TextIO.sink())
+                    .to(outputBucket())
+                    .withNaming(key -> new CsvFileNaming(key, ".csv"))
+                    .withNumShards(1));
+
+    }
+
+    public static class CsvFileNaming implements FileIO.Write.FileNaming {
+        private String fileName;
+        private String suffix;
+
+        public CsvFileNaming(String fileName, String suffix) {
+            this.fileName = fileName;
+            this.suffix = suffix;
+        }
+
+        @Override
+        public @UnknownKeyFor @NonNull @Initialized String getFilename(@UnknownKeyFor @NonNull @Initialized BoundedWindow window, @UnknownKeyFor @NonNull @Initialized PaneInfo pane, @UnknownKeyFor @NonNull @Initialized int numShards, @UnknownKeyFor @NonNull @Initialized int shardIndex, @UnknownKeyFor @NonNull @Initialized Compression compression) {
```

Comment:
What is the reason behind naming file with shards when number of shards is always fixed to 1? Also is there a reason behind not to keep the filename in sink same as input file name?